File: 1106-parsing-a-boolean-expression/1106-parsing-a-boolean-expression.cpp
class Solution {
public:
    bool parseBoolExpr(string expression) {
        int s = 0;
        return helper(expression, s);
    }
        
    bool helper(string e, int& s) {
        char ch = e[s++];
        if(ch == 't') return true;
        if(ch == 'f') return false;
        
        if(ch == '!') {
            bool ans = !helper(e, ++s);
            ++s;
            return ans;
        }
        
        bool is_and = (ch == '&');
        bool ans = is_and;
        ++s;
        while(true) {
            if (is_and) ans &= helper(e, s);
            else ans |= helper(e, s);
            if (e[s++] == ')') break;
        }

        return ans;

    }
};

File: 1106-parsing-a-boolean-expression/1106-parsing-a-boolean-expression.java
class Solution {
    int s = 0;
    char[] exp;
    public boolean parseBoolExpr(String expression) {
        exp = expression.toCharArray();
        return parse() == 1;
    }
    
    int parse() {
        char ch = exp[s++];
        if(ch == 't') return 1;
        if(ch == 'f') return 0;
        if(ch == '!') {
            ++s;
            int ans = parse();
            ++s;
            return 1 - ans;
        }
        boolean isAnd = (ch == '&');
        int ans = isAnd ? 1 : 0;
        ++s;
        while(true) {
            if(isAnd) ans &= parse();
            else ans |= parse();
            if(exp[s++] == ')') break;
        }
        return ans;
    }
}